File: library/CreateDataset.js
const fs = require('fs');

let create_dataset = function(endTime="2020-05-04",predTime=null){
  t0 = new Date("2019-12-31");
  tf = new Date(endTime);

  // read all rows and chaeck the border time and convert the selected colnames
	let temp, file;
	file = fs.readFileSync('../samples/ON.csv').toString();
	let lines = file.split(/\r\n|\n/);
	let dataReport = [];

  delay = 0;
  let startdate = new Date("2020-01-01");
	for (let i = 1; i < lines.length; i++) {
		let temp = lines[i].split(',');
		if(temp.length > 1) {
			tempDate = new Date(temp[0]);
			if( tempDate > startdate && tempDate <= tf - delay){
				let tempDayCount = Math.ceil(Math.abs(tempDate - t0) / (1000 * 3600 * 24));
				dataReport.push([tempDayCount, Number(temp[1])]);
			}
		}
  }
  
	let timeLength = dataReport[dataReport.length-1][0];
	let time = Array.from(Array(timeLength), (_, i) => i + 1);
  
  reports = Array.from(Array(timeLength)).fill(0);
  for(let i = 0; i < dataReport.length; i++)
    reports[dataReport[i][0] - 1] = dataReport[i][1];
  for(let i = timeLength - delay - 1; i < timeLength; i++)
    reports[i] = NaN;

  // read all rows and chaeck the border time and convert the selected colnames
	temp, file;
	file = fs.readFileSync('../samples/covidtesting.csv').toString();
	lines = file.split(/\r\n|\n/);
  let selected_colnames = ["Reported Date",
    "Deaths",
    "Number of patients hospitalized with COVID-19",
    "Number of patients in ICU with COVID-19",
    "Number of patients in ICU on a ventilator with COVID-19"];
	let dataCovar_Index = [];	
	let dataCovar = [];
	let dataCovar_temp = lines[0].replace(/['"]+/g, '').split(',');
	for(let i = 0; i < selected_colnames.length; i++){
		dataCovar_Index.push(dataCovar_temp.indexOf(selected_colnames[i]));
	}
  startdate = new Date("2020-03-17");
	for (let i = 1; i < lines.length; i++) {
		let temp = lines[i].split(',');
		if(temp.length > 1) {
			tempDate = new Date(temp[dataCovar_Index[0]]);
			if( tempDate > startdate && tempDate <= tf){
				let tempDayCount = Math.ceil(Math.abs(tempDate - t0) / (1000 * 3600 * 24));
			  let tempcovar =	[tempDayCount];
				for(let j = 1; j < selected_colnames.length; j++){
					tempcovar.push(temp[dataCovar_Index[j]] == '' ? NaN :Number(temp[dataCovar_Index[j]]));
				}
				dataCovar.push(tempcovar);
			}
		}
  }
  
  total_deaths = Array.from(Array(timeLength)).fill(0);
  hospital = Array.from(Array(timeLength)).fill(NaN);
  ICU = Array.from(Array(timeLength)).fill(NaN);
  ventilator = Array.from(Array(timeLength)).fill(NaN);
  for(let i = 0; i < dataCovar.length; i++){
    total_deaths[dataCovar[i][0] - 1] = dataCovar[i][1];
    hospital[dataCovar[i][0] - 1] = dataCovar[i][2];
    ICU[dataCovar[i][0] - 1] = dataCovar[i][3];
    ventilator[dataCovar[i][0] - 1] = dataCovar[i][4];
  }
  deaths = total_deaths.map( (_,i,arr) => (arr[i]) - (arr[i-1]));
	deaths[0] = total_deaths[0];
	
  let data = time.map( (x,i) => [x, reports[i], deaths[i], hospital[i], ICU[i], ventilator[i]]);

  for(let i = 0; i < 23 - 1; i++){
    data[i][3] = 0;
    data[i][4] = 0;
    data[i][5] = 0;
  }
  
  data.unshift(["time","reports","deaths","hospital","ICU","ventilator"]);

	if (predTime != null) {
    predTime = new Date(predTime);
		let add_time = Math.ceil(Math.abs(tf - predTime) / (1000 * 3600 * 24));
		for(let i =0; i < add_time; i++)
			data.push([timeLength + i, NaN, NaN, NaN, NaN, NaN]);
	}
	return data;
}

module.exports = create_dataset;

File: R/CreateDataset.R

create_dataset <- function(endTime="2020-05-04",predTime=NULL) {

  t0 <- as.Date("2019-12-31")
  tf <- as.Date(endTime)
  dat <- read.csv("../samples/ON.csv", header=TRUE)
  dat$Date <- as.Date(dat$Date)

  delay <- 0
  dat <- subset(dat, Date>as.Date("2020-01-01")&Date<=tf-delay)
  dat$Date <- as.numeric(dat$Dat)
  dat$Date <- dat$Date - as.numeric(t0)

  time <- seq(1,as.numeric(tf)-as.numeric(t0),by=1)
  reports <- rep(0,length(time))
  reports[dat$Date] <- dat$Cases
  reports[(length(time)-delay):length(time)] <- NA


  dat <- read.csv("../samples/covidtesting.csv", header=TRUE)
  dat$Reported.Date <- as.Date(dat$Reported.Date)
  dat <- subset(dat, Reported.Date>as.Date("2020-03-17") & Reported.Date<=tf)
  dat$Reported.Date <- as.numeric(as.Date(dat$Reported.Date))-as.numeric(t0)
  dat <- subset(dat, select=c("Reported.Date","Deaths",
                              "Number.of.patients.hospitalized.with.COVID.19",
                              "Number.of.patients.in.ICU.with.COVID.19",
                              "Number.of.patients.in.ICU.on.a.ventilator.with.COVID.19"))
  colnames(dat) <- c("Date","Deaths","Hospital","ICU","Ventilator")

  total_deaths <- rep(0,length(time))
  total_deaths[dat$Date] <- dat$Deaths
  deaths <- rep(0,length(time))

  for (i in 2:length(time)) {
    deaths[i] <- total_deaths[i]-total_deaths[i-1]
  }

  hospital <- rep(NA,length(time))
  hospital[dat$Date] <- dat$Hospital

  ICU <- rep(NA,length(time))
  ICU[dat$Date] <- dat$ICU

  ventilator <- rep(NA,length(time))
  ventilator[dat$Date] <- dat$Ventilator

  data <- data.frame(time,reports,deaths,hospital,ICU,ventilator)
  colnames(data) <- c("time","reports","deaths","hospital","ICU","ventilator")

  ind <- which(data$time < 23)
  data[ind,c("hospital","ICU","ventilator")] <- 0

  if (!is.null(predTime)) {
    predTime <- as.Date(predTime)
    add_time <- seq(as.numeric(tf)+1,as.numeric(as.Date(predTime)))-as.numeric(t0)
    add_reports <- rep(NA,length(add_time))
    add_deaths <- rep(NA,length(add_time))
    add_hospital <- rep(NA,length(add_time))
    add_ICU <- rep(NA,length(add_time))
    add_ventilator <- rep(NA,length(add_time))
    add_data <- cbind(add_time,add_reports,add_deaths,add_hospital,add_ICU,add_ventilator)
    colnames(add_data) <- c("time","reports","deaths","hospital","ICU","ventilator")
    data <- rbind(data,add_data)
    data <- as.data.frame(data)

  }

  return(data)
}
